formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/DynamoUpdateService.scala

```diff
@@ -0,0 +1,81 @@
+package com.gu.identity.formstackbatonrequests
+
+import java.time.{Instant, LocalDate, LocalDateTime}
+
+import com.gu.identity.formstackbatonrequests.aws.{DynamoClient, SubmissionTableUpdateDate}
+import com.gu.identity.formstackbatonrequests.circeCodecs.{Form, FormSubmission}
+import com.typesafe.scalalogging.LazyLogging
+
+case class DynamoUpdateService(
+  formstackClient: FormstackRequestService,
+  dynamoClient: DynamoClient,
+  config: PerformLambdaConfig) extends LazyLogging {
+
+  def submissionsWithEmailAndAccount(submissions: List[FormSubmission], accountNumber: Int): List[SubmissionIdEmail] = {
+    val emailReg = """(?i)\b[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}\b""".r
+    submissions.foldLeft(List.empty[SubmissionIdEmail]) { (acc, submission) =>
+      val submissionValues = submission.data.map(field => field._2.value).toList
+      val emailList = submissionValues.collect { case jsonValue => emailReg.findAllIn(jsonValue.toString).toList }.flatten
+      val receivedByLambdaTimestamp = Instant.now.getEpochSecond
+      val submissionsIdEmails = emailList.map(email => SubmissionIdEmail(email.toLowerCase, submission.id, receivedByLambdaTimestamp, accountNumber))
+      submissionsIdEmails ::: acc
+    }
+  }
+
+  private def skipDecryptionError(err: Throwable): Either[Throwable, Unit] = {
+    err match {
+      case _: FormstackDecryptionError => Right(())
+      case err => Left(err)
+    }
+  }
+
+  private def writeSubmissions(form: Form, lastUpdate: SubmissionTableUpdateDate, submissionPage: Int = 1, token: FormstackAccountToken): Either[Throwable, Unit] = {
+    formstackClient.formSubmissionsForGivenPage(submissionPage, form.id, lastUpdate, config.encryptionPassword, token) match {
+      case Left(err) => skipDecryptionError(err)
+      case Right(response) =>
+        logger.info(s"Received page $submissionPage of submissions out of ${response.pages} pages for form ${form.id}.")
+        val submissionsIdsWithEmails = submissionsWithEmailAndAccount(response.submissions, token.account)
+        logger.info(s"Writing ${submissionsIdsWithEmails.length} submission id(s) and emails to Dynamo")
+        dynamoClient.writeSubmissions(submissionsIdsWithEmails, config.bcryptSalt, config.submissionTableName) match {
```

Comment:
may be creating 2-3 small functions here would make this more readable. too many right left in this function.

Reply:
Yeah.. I'm struggled to find a good was of breaking this down, since it's recursive and it would mean passing so many parameters. Does anything obvious jump out to you?